airbyte-cdk/python/airbyte_cdk/sources/streams/concurrent/cursor.py

```diff
@@ -0,0 +1,153 @@
+import functools
+from abc import ABC, abstractmethod
+from typing import Any, List, Mapping, Optional, Protocol, Tuple
+
+from airbyte_cdk.sources.connector_state_manager import ConnectorStateManager
+from airbyte_cdk.sources.message import MessageRepository
+from airbyte_cdk.sources.streams.concurrent.partitions.partition import Partition
+from airbyte_cdk.sources.streams.concurrent.partitions.record import Record
+
+
+def _extract_value(mapping: Mapping[str, Any], path: List[str]) -> Any:
+    return functools.reduce(lambda a, b: a[b], path, mapping)
+
+
+class Comparable(Protocol):
+    """Protocol for annotating comparable types."""
+
+    @abstractmethod
+    def __lt__(self: "Comparable", other: "Comparable") -> bool:
+        pass
+
+
+class CursorField:
+    def __init__(self, path: List[str]) -> None:
+        self._path = path
+
+    def extract_value(self, record: Record) -> Comparable:
+        return _extract_value(record.data, self._path)  # type: ignore  # we assume that the value the path points at is a comparable
+
+
+class Cursor(ABC):
+    @abstractmethod
+    def observe(self, record: Record) -> None:
+        """
+        Indicate to the cursor that the record has been emitted
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def close_partition(self, partition: Partition) -> None:
+        """
+        Indicate to the cursor that the partition has been successfully processed
+        """
+        raise NotImplementedError()
+
+    @abstractmethod
+    def end_sync(self) -> None:
+        """
+        Indicate to the cursor that the sync has been successfully processed
+        """
+        raise NotImplementedError()
+
+
+class NoopCursor(Cursor):
+    def observe(self, record: Record) -> None:
+        pass
+
+    def close_partition(self, partition: Partition) -> None:
+        pass
+
+    def end_sync(self) -> None:
+        pass
+
+
+class ConcurrentCursor(Cursor):
+    _START_BOUNDARY = 0
+    _END_BOUNDARY = 1
+
+    def __init__(
+        self,
+        stream_name: str,
+        stream_namespace: Optional[str],
+        stream_state: Any,
```

Comment:
can't we restrict stream_state to `Mapping[str, Any]`?

Reply:
Probably. I haven't taken time to see what it would look like when we'll actually use this variable. I typed `Any` here because until migration is consider, this piece of code didn't care about the typing